docs/proposals/trimming-friendly-registrar-proposal.md

```diff
@@ -0,0 +1,483 @@
+
+### Option C: Source generator + post-trimming MSBuild task
+
+The code for the UCO callbacks should be expressible in C#. The only problem that we might have to work around is the
+absence of the "inverse dependency" attributes. It might be a good use case to drive the implementation of this feature.
+
+The post-trimming task would be exactly the same as in Option B.
+
+- **Pros**:
+  - likely faster than an IL-weaving MSBuild task
+    - code is generated while editing the source files
+    - the code for the xamarin runtime and for all libraries (e.g, MAUI) would already be pre-built in managed DLLs
+  - code generator can produce errors and warnings while editing the source files for better DX
+  - source generators are more future-proof than Mono.Cecil
+- **Cons**:
+  - the generated code has to be distributed with runtimepack/nupkg
+    - all libraries authors will have to publish their code specifically for .NET 9 and any future releases
+      which change the registrar
+  - generating Objective-C code in roslyn source generators is a _hack_ that isn't officially supported
+    - source generators are _synchronous_ and doing many IO operations will slow them down, causing lagging in the editor
+  - its's not yet clear how to correctly generate the `[DynamicDependency]` attributes and if the inverse dependency
+    attributes could be added to .NET
+
+### Notes
+
+- I suggest implementing the change with the Mono.Cecil-based approach since it should be much faster to implement and
+  overall less risky.
+- The C# source generator is a more tempting option and probably more future-proof.
+  - I'm worried it could take a lot longer to implement and there might have problems with third party libraries.
+- This registrar relies on static linking and doesn't support the non-AOT scenario (using `xamarin_registrar_dlsym` to
+  load function pointers to the registrar callbacks).
+    - This is not a problem in the context of NativeAOT but it might mean that the registrar could not completely replace
+      the existing MSR and static registrar for Mono-based (CoreCLR-based?) scenarios and there will still be need
+      to keep the custom linker steps around to build those registrars.
+- There are certainly edge-cases that aren't covered in this document but they should all be possible to implement
+  using the same ideas that are outlined in this document.
+  - This claim needs validation though. Are there some obvious cases which would be hard or obviously inefficient
+    to implement this way?
+- It is unclear what would be the file size and startup or runtime performance implications of these changes.
+  - We will need to generate less managed code but on the other hand we'll generate more Objective-C code.
+  - There will be more transitions between managed and native code. Does it matter too much in the context of NativeAOT
+    though? Will it be worse than performing linear search in the lookup tables that we do today?
```

Comment:
This is likely app-dependant: a linear search in a small table is quick, while in a bigger table it gets quite slow. This new registrar has constant speed, and there are probably optimizations that can be made too if we find bottlenecks.